pkg/registry/apis/query/parser.go

```diff
@@ -1,83 +1,209 @@
+
+	// Headers that should be forwarded to the next request
+	Headers map[string]string `json:"headers,omitempty"`
 }
 
-type groupedQueries struct {
-	// the plugin type
-	pluginId string
+type parsedRequestInfo struct {
+	// Datasource queries, one for each datasource
+	Requests []datasourceRequest `json:"requests,omitempty"`
 
-	// The datasource name/uid
-	uid string
+	// Expressions in required execution order
+	Expressions []expr.ExpressionQuery `json:"expressions,omitempty"`
 
-	// The raw backend query objects
-	query []v0alpha1.GenericDataQuery
+	// Expressions include explicit hacks for influx+prometheus
+	RefIDTypes map[string]string `json:"types,omitempty"`
+
+	// Hidden queries used as dependencies
+	HideBeforeReturn []string `json:"hide,omitempty"`
 }
 
-// Internally define what makes this request unique (eventually may include the apiVersion)
-func (d *groupedQueries) key() string {
-	return fmt.Sprintf("%s/%s", d.pluginId, d.uid)
+type queryParser struct {
+	legacy service.LegacyDataSourceLookup
+	reader *expr.ExpressionQueryReader
+	tracer tracing.Tracer
 }
 
-func parseQueryRequest(raw v0alpha1.GenericQueryRequest) (parsedQueryRequest, error) {
-	mixed := make(map[string]*groupedQueries)
-	parsed := parsedQueryRequest{}
-	refIds := make(map[string]bool)
+func newQueryParser(reader *expr.ExpressionQueryReader, legacy service.LegacyDataSourceLookup, tracer tracing.Tracer) *queryParser {
+	return &queryParser{
+		reader: reader,
+		legacy: legacy,
+		tracer: tracer,
+	}
+}
+
+// Split the main query into multiple
+func (p *queryParser) parseRequest(ctx context.Context, input *query.QueryDataRequest) (parsedRequestInfo, error) {
+	ctx, span := p.tracer.Start(ctx, "QueryService.parseRequest")
+	defer span.End()
+
+	queryRefIDs := make(map[string]*data.DataQuery, len(input.Queries))
+	expressions := make(map[string]*expr.ExpressionQuery)
+	index := make(map[string]int) // index lookup
+	rsp := parsedRequestInfo{
+		RefIDTypes: make(map[string]string, len(input.Queries)),
+	}
+
+	// Ensure a valid time range
+	if input.From == "" {
+		input.From = "now-6h"
+	}
+	if input.To == "" {
+		input.To = ""
+	}
 
-	for _, original := range raw.Queries {
-		if refIds[original.RefID] {
-			return parsed, fmt.Errorf("invalid query, duplicate refId: " + original.RefID)
+	for _, q := range input.Queries {
+		_, found := queryRefIDs[q.RefID]
+		if found {
+			return rsp, fmt.Errorf("multiple queries found for refId: %s", q.RefID)
+		}
+		_, found = expressions[q.RefID]
+		if found {
+			return rsp, fmt.Errorf("multiple queries found for refId: %s", q.RefID)
 		}
 
-		refIds[original.RefID] = true
-		q := original
+		ds, err := p.getValidDataSourceRef(ctx, q.Datasource, q.DatasourceID)
+		if err != nil {
+			return rsp, err
+		}
 
-		if q.TimeRange == nil && raw.From != "" {
-			q.TimeRange = &v0alpha1.TimeRange{
-				From: raw.From,
-				To:   raw.To,
+		// Process each query
+		if expr.IsDataSource(ds.UID) {
+			raw, err := json.Marshal(q)
```

Comment:
could you add a comment here explaining why it's going from object -> bytes -> iter -> exp object?